---
title: "Transcriptome analysis manual"
author: "Varvara Tvorogova"
date: "3 April 2021"
output: html_document
---
This manual describes common transcriptome processing for differential expression analysis

First of all, you need to download files with raw reads with .fastq. They are usually packed with gunzip and have .fastq.gz extention. Move them to the new folder, which contains only fastq.gz files analysed.
Now you need to evaluate the quality of your reads. You can use fastqc program for that.
Fastqc can be downloaded from here https://www.bioinformatics.babraham.ac.uk/projects/download.html
You do not need to install it, just unpack it in some directory.
To run fastqc

```{bash}
for FQ in 
do

```

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r cars}
summary(cars)
```

## Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
